modules/ballerina-core/src/main/java/org/ballerinalang/helpers/DeepEqualHelper.java

```diff
@@ -0,0 +1,279 @@
+                    return false;
+                }
+    
+                List<String> keys = Arrays.stream(lhMap.keySet().toArray())
+                        .map(String.class::cast)
+                        .collect(Collectors.toList());
+                for (int i = 0; i < lhMap.size(); i++) {
+                    if (!isDeepEqual(lhMap.get(keys.get(i)), rhMap.get(keys.get(i)))) {
+                        return false;
+                    }
+                }
+                break;
+            case TypeTags.ARRAY_TAG:
+            case TypeTags.ANY_TAG:
+                // TODO: This block should ideally be in the ARRAY_TAG case only. Not ANY_TAG. #4505.
+                if (lhVal instanceof BNewArray) {
+                    BNewArray lhArray = (BNewArray) lhVal;
+                    BNewArray rhArray = (BNewArray) rhVal;
+                    if (lhArray.size() != rhArray.size()) {
+                        return false;
+                    }
+    
+                    for (int i = 0; i < lhArray.size(); i++) {
+                        if (!isDeepEqual(lhArray.getBValue(i), rhArray.getBValue(i))) {
+                            return false;
+                        }
+                    }
+                }
+                
+                break;
+            case TypeTags.JSON_TAG:
+                JsonNode lhJSON = ((BJSON) lhVal).value();
+                JsonNode rhJSON = ((BJSON) rhVal).value();
+    
+                return isDeepEqual(lhJSON, rhJSON);
+            case TypeTags.XML_TAG:
+                // https://www.pixelstech.net/article/1453272563-Canonicalize-XML-in-Java
+                break;
+            case TypeTags.CONNECTOR_TAG:
+                BConnector lhConnector = (BConnector) lhVal;
+                BConnector rhConnector = (BConnector) rhVal;
+                
+                if (!lhConnector.getConnectorType().getPackagePath().equals(
+                        rhConnector.getConnectorType().getPackagePath())) {
+                    return false;
+                }
+    
+                if (!lhConnector.getConnectorType().getName().equals(rhConnector.getConnectorType().getName())) {
+                    return false;
+                }
+                break;
+            case TypeTags.ENUM_TAG:
+                BEnumerator lhEnum = (BEnumerator) lhVal;
+                BEnumerator rhEnum = (BEnumerator) rhVal;
+                
+                if (!lhEnum.stringValue().equals(rhEnum.stringValue())) {
+                    return false;
+                }
+                break;
+            default:
+                return false;
+        }
+        
+        return true;
+    }
+    
+    /**
+     * Deep equals function for JSON type values. I.E ===
+     * @param lhJson The left side value.
+     * @param rhJson The right side value.
+     * @return True if values are equal, else false.
+     */
+    public static boolean isDeepEqual(JsonNode lhJson, JsonNode rhJson) {
```

Comment:
We can make this method private.

Reply:
Fixed with https://github.com/ballerina-lang/ballerina/pull/4532/commits/c728619b2fcdb290f1587e71e11911955ccfae3f